conn.go

```diff
@@ -1255,67 +1271,60 @@ func (c *Conn) doRequest(d *connDeadline, write func(time.Time, int32) error) (i
 		// recoverable state so we're better off just giving up at this point to
 		// avoid any risk of corrupting the following operations.
 		c.conn.Close()
+		c.leave()
 	}
 
 	c.wlock.Unlock()
 	return
 }
 
 func (c *Conn) waitResponse(d *connDeadline, id int32) (deadline time.Time, size int, lock *sync.Mutex, err error) {
-	// I applied exactly zero scientific process to choose this value,
-	// it seemed to worked fine in practice tho.
-	//
-	// My guess is 100 iterations where the goroutine gets descheduled
-	// by calling runtime.Gosched() may end up on a wait of ~10ms to ~1s
-	// (if the programs is heavily CPU bound and has lots of goroutines),
-	// so it should allow for bailing quickly without taking too much risk
-	// to get false positives.
-	const maxAttempts = 100
-	var lastID int32
-
-	for attempt := 0; attempt < maxAttempts; {
+	c.rlock.Lock()
+
+	for {
 		var rsz int32
 		var rid int32
 
-		c.rlock.Lock()
 		deadline = d.setConnReadDeadline(c.conn)
 
 		if rsz, rid, err = c.peekResponseSizeAndID(); err != nil {
-			d.unsetConnReadDeadline()
-			c.conn.Close()
-			c.rlock.Unlock()
-			return
+			break
 		}
 
 		if id == rid {
 			c.skipResponseSizeAndID()
 			size, lock = int(rsz-4), &c.rlock
-			return
+			break
+		}
+
+		if c.concurrency() == 1 {
+			// If the goroutine is the only one waiting on this connection it
+			// should be impossible to read a correlation id different from the
+			// one it expects. This is a sign that the data we are reading on
+			// the wire is corrupted and the connection needs to be closed.
+			err = io.ErrNoProgress
+			break
 		}
 
 		// Optimistically release the read lock if a response has already
 		// been received but the current operation is not the target for it.
 		c.rlock.Unlock()
 		runtime.Gosched()
+		c.rlock.Lock()
```

Comment:
I think it would read a little more clearly if you added a `c.rlock.Lock()` as the first statement in the loop and removed this one and the one before entering the loop.  IIUC, the logic would be the same.

Reply:
I tried to refactor the code to apply this suggestion, but I think it's actually making it harder to understand: the lock is not always released, sometimes it's returned to the caller to transfer ownership, so having it locked in the loop makes this case (which is the common case) much harder to reason about, so I think we should keep it as it is for now.

Reply:
Actually after re-reading your other comment I think I have something that's not too tricky, let me know what you think about it.

Reply:
Nice...I like that a lot more. 😁